### PR TITLE
Discover network peers using DID service

### LINKED ISSUE #105: VDR/Network: discover new peers from DID service

### PR DESCRIPTION
**Pull request by reinkrul**
Fixes #105

### CODE REVIEW

File: core/echo_mock.go [hunk 1]
```diff
@@ -5,35 +5,36 @@
...
 import (
+	reflect "reflect"
+
 	gomock "github.com/golang/mock/gomock"
 	echo "github.com/labstack/echo/v4"
-	reflect "reflect"
 )
 
-// MockEchoServer is a mock of EchoServer interface
+// MockEchoServer is a mock of EchoServer interface.
```
Comment (woutslakhorst) [inline core/echo_mock.go:14]: did you install the latest tools using the makefile? If so, what version is your mockgen on?

File: docs/pages/did-administration-cookbook.rst [hunk 2]
```diff
@@ -0,0 +1,18 @@
+.. _did-administration-cookbook:
+
+DID Document Administration Cookbook
```
Comment (woutslakhorst) [inline docs/pages/did-administration-cookbook.rst:3]: this page won't be added to the index.

File: network/interface.go [hunk 3]
```diff
@@ -41,4 +41,7 @@ type Transactions interface {
 	CreateDocument(payloadType string, payload []byte, signingKeyID string, attachKey crypto2.PublicKey, timestamp time.Time, fieldsOpts ...dag.FieldOpt) (dag.Document, error)
 	// ListDocuments returns all documents known to this Network instance.
 	ListDocuments() ([]dag.Document, error)
+	// AddPeer instructs the P2P layer to try to connect to a new peer on the given address. It's safe to call it
+	// multiple times for the same address. If the P2P layer will attempt to connect to the address it returns `true`.
+	AddPeer(address string) bool
```
Comment (woutslakhorst) [inline network/interface.go:46]: Is there also a remove peer? eg: when an endpoint is removed...

File: vdr/ambassador.go [hunk 4]
```diff
@@ -335,3 +341,20 @@ func (n ambassador) findKeyByThumbprint(thumbPrint []byte, didDocumentAuthKeys [
 	}
 	return documentKey, nil
 }
+
+// discoverPeers tries to find Nuts Network services in the given DID Document to add them as Nuts Network peers.
+func (n ambassador) discoverPeers(document did.Document, _ types.DocumentMetadata) error {
+	for _, service := range document.Service {
+		if service.Type == "nuts-network-grpc" {
```
Comment (woutslakhorst) [inline vdr/ambassador.go:348]: constant

Comment (woutslakhorst) [reply vdr/ambassador.go:348]: our docs should reflect this feature and enable admins to change the endpoint. If part of future development, it must be stated in an issue.

File: vdr/types/common.go [hunk 5]
```diff
@@ -70,3 +71,18 @@ type ResolveMetadata struct {
...
 	AllowDeactivated bool
 }
+
+// DocWriterDecorator wraps a DocWriter with a function that's called before the target writer is called.
+type DocWriterDecorator struct {
+	Target    DocWriter
+	Decorator func(document did.Document, metadata DocumentMetadata) error
+}
+
+func (d DocWriterDecorator) Write(document did.Document, metadata DocumentMetadata) error {
```
Comment (woutslakhorst) [inline vdr/types/common.go:81]: If the decorator fails, nothing is written. If the docWriter fails, connection will still be made....
The ordering and validity is very important and must be put in our specifications first.
If the decorator pattern is valid, depends on the specs.

**Review by woutslakhorst — CHANGES_REQUESTED** [review]
This PR feels very quick and dirty. It's part of a bigger design including:

- service specification in DIDs, what happens if the endpoint is incorrect for the given type. This is mainly part of a bolt, but in this case it's not.
- what if endpoints are removed?
- what if nodes are down (and will never come back)
- documentation
- helper APIs? Since manipulating endpoints is quite hard right now.

An epic issue can be used to track the entire progress.

**Comment by reinkrul** [comment]
> This PR feels very quick and dirty. It's part of a bigger design including:
> 
> * service specification in DIDs, what happens if the endpoint is incorrect for the given type. This is mainly part of a bolt, but in this case it's not.
> * what if endpoints are removed?
> * what if nodes are down (and will never come back)
> * documentation
> * helper APIs? Since manipulating endpoints is quite hard right now.
> 
> An epic issue can be used to track the entire progress.

I tried to spec it in https://github.com/nuts-foundation/nuts-specification/pull/34

But: since this I remember discussing this I thought we could just squeeze it in but it has more facets to it than I anticipated. Since we don't really need peer discovery at this point because there aren't that many peers to begin with (which can be configured perfectly fine as bootstrap nodes).

I'll keep the spec PR so we can work on it, but I'll close this one for now (for later work).

**Pull request closed without merging**